internal/dag/gatewayapi_processor.go

```diff
@@ -195,39 +196,121 @@ func isSecretRef(certificateRef *gatewayapi_v1alpha1.LocalObjectReference) bool
+	if len(hostnames) == 0 && listenerHostname == nil {
 		hosts = append(hosts, "*")
 		return hosts, nil
 	}
 
-	for _, host := range route.Spec.Hostnames {
+	if listenerHostname != nil {
+		if string(*listenerHostname) != "*" {
+
+			// Validate listener hostname.
+			if err := validHostName(string(*listenerHostname)); err != nil {
+				return hosts, []error{err}
+			}
+
+			if len(hostnames) == 0 {
+				hosts = append(hosts, string(*listenerHostname))
+				return hosts, nil
+			}
+		}
+	}
+
+	for _, host := range hostnames {
 
 		hostname := string(host)
-		if isIP := net.ParseIP(hostname) != nil; isIP {
-			errors = append(errors, fmt.Errorf("hostname %q must be a DNS name, not an IP address", hostname))
+
+		// A "*" hostname matches anything.
+		if hostname == "*" {
+			hosts = append(hosts, hostname)
 			continue
 		}
-		if strings.Contains(hostname, "*") {
-			if errs := validation.IsWildcardDNS1123Subdomain(hostname); errs != nil {
-				errors = append(errors, fmt.Errorf("invalid hostname %q: %v", hostname, errs))
+
+		// Validate the hostname.
+		if err := validHostName(hostname); err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		if listenerHostname != nil {
+			const wildcardString string = "[a-z0-9]([-a-z0-9]*[a-z0-9])*."
+			lhn := string(*listenerHostname)
+
+			// A "*" hostname matches anything.
+			if lhn == "*" {
+				hosts = append(hosts, hostname)
 				continue
 			}
-		} else {
-			if errs := validation.IsDNS1123Subdomain(hostname); errs != nil {
-				errors = append(errors, fmt.Errorf("invalid listener hostname %q: %v", hostname, errs))
+
+			// If the listener.hostname matches then no need to
+			// do any other validation.
+			if lhn == hostname {
+				hosts = append(hosts, hostname)
 				continue
 			}
+
+			if strings.Contains(lhn, "*") {
+
+				// Build a regex to match the hostname against the wildcard listener hostname.
+				wildcardDNS1123SubdomainRegexp := regexp.MustCompile("^" + strings.Replace(lhn, "*", wildcardString, 1) + "$")
+
+				// Validate the wildcard gateway listener hostname matches the hostnames hostname.
+				if !wildcardDNS1123SubdomainRegexp.MatchString(hostname) {
+					errors = append(errors, fmt.Errorf("gateway hostname %q does not match hostnames hostname %q", lhn, hostname))
+					continue
+				}
+			} else if strings.Contains(hostname, "*") {
+
+				// Build a regex to match the hostname against the wildcard hostnames hostname.
+				wildcardDNS1123SubdomainRegexp := regexp.MustCompile("^" + strings.Replace(hostname, "*", wildcardString, 1) + "$")
+
+				// Validate the wildcard hostnames hostname matches gateway listener hostname.
+				if !wildcardDNS1123SubdomainRegexp.MatchString(lhn) {
+					errors = append(errors, fmt.Errorf("hostnames hostname %q does not match gateway hostname %q", hostname, lhn))
+					continue
+				}
+			} else {
+				// Validate the gateway listener hostname matches the hostnames hostname.
+				if lhn != hostname {
```

Comment:
I think this check for non-equality is redundant/can be removed since we already checked for equality on https://github.com/projectcontour/contour/pull/3640/files#diff-2dfb3932d84ad36571b20cb8d66d9f0b623fdbbef49a0b41eff48eaa7fa22a05R258

Reply:
We need this to cover this test case where the hostnames don't match without wildcards: https://github.com/projectcontour/contour/blob/d0ed425ce3a3d6950e989e3ae58963fea5e2bfae/internal/dag/gatewayapi_processor_test.go#L119

Reply:
We need the `else`, but since we already checked for `if lhn == hostname` above, then we know if we got here that `lhn != hostname`. It's not a big deal, can leave.

Reply:
Ohh I see now, I thought you meant to remove the `if` statement. I updated. =)